### PR TITLE
buf: add module

### PR DESCRIPTION
**Pull request by lucperkins**
### Description

This PR adds a new module for the [Buf CLI tool](https://github.com/bufbuild/buf), a tool for working with Protocol Buffers. This is my first time contributing here so go easy on me 😄

I've included a test for the module, which passes:

```sh
nix-shell --pure tests -A run.buf
...
buf: OK
```

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/buf.nix [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+with lib;
+
+let
+  cfg = config.programs.buf;
+in {
+  meta.maintainers = [ maintainers.lucperkins ];
+
+  options.programs.buf = {
+    enable = mkEnableOption "The Buf CLI tool";
```
Comment (mainrs) [inline modules/programs/buf.nix:11]: Not sure if there is a set convention here. But a lot of packages seem to use `buf, a protocol buffer CLI` style names. 
Then again, a lot don't.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -68,6 +68,8 @@
 
 /modules/programs/broot.nix                           @aheaume
 
+/modules/programs/buf.nix                             @lucperkins
+
```
Comment (mainrs) [inline .github/CODEOWNERS:72]:
```suggestion
/modules/programs/buf.nix                             @lucperkins
/tests/modules/programs/buf                           @lucperkins

```

This is definitely common to do!

**Review by mainrs — CHANGES_REQUESTED** [review]

**Comment by mainrs** [comment]
Note I am not a maintainer, it's just stuff I noticed :)

File: .github/CODEOWNERS (hunk 2 above)
Comment (lucperkins) [reply .github/CODEOWNERS:72]: Ah yes! Did not catch that 😄

File: modules/programs/buf.nix (hunk 1 above)
Comment (lucperkins) [reply modules/programs/buf.nix]: I opted for something a bit more descriptive

File: modules/programs/buf.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      '';
+    };
+
+    enableZshIntegration = mkOption {
+      default = false;
+      type = types.bool;
+      description = ''
+        Whether to load zsh autocomplete scripts for <command>buf</command>.
+      '';
+    };
```
Comment (rycee) [inline modules/programs/buf.nix:41]: Aren't the completions picked up automatically? Looking at the `buf` package in Nixpkgs it does seem in generate and install the completions: https://github.com/NixOS/nixpkgs/blob/78bdd3bd437281896695e3947a81fa5609e9548c/pkgs/development/tools/buf/default.nix#L56-L60

Comment (lucperkins) [reply modules/programs/buf.nix]: Ah yes, you are right. I've removed these options.

File: modules/programs/buf.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let cfg = config.programs.buf;
+in {
+  meta.maintainers = [ maintainers.lucperkins ];
+
+  options.programs.buf = {
+    enable = mkEnableOption "Buf: a CLI tool for working with Protocol Buffers";
```
Comment (sumnerevans) [inline modules/programs/buf.nix:10]: note that whatever you put inside this option will be surrounded by `Whether to enable {text}.`, so I think a comma makes more sense. Also, I don't think that "Protocol Buffers" is a proper noun:

```suggestion
    enable = mkEnableOption "Buf, a CLI tool for working with protocol buffers";
```

File: modules/programs/buf.nix [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.buf;
+      defaultText = literalExpression "pkgs.buf";
+      description = "Package providing <command>buf</command>.";
+    };
+  };
+
+  config = mkIf cfg.enable {
```
Comment (sumnerevans) [inline modules/programs/buf.nix:20]: Not sure how much value this package is offering as is. Is there any config for this tool that we can provide options for?

File: modules/programs/buf.nix (hunk 4 above)
Comment (lucperkins) [reply modules/programs/buf.nix:10]: There is some variation but in general it's capitalized. See:

https://en.wikipedia.org/wiki/Protocol_Buffers
https://github.com/protocolbuffers/protobuf
https://cloud.google.com/apis/design/proto3

File: modules/programs/buf.nix (hunk 5 above)
Comment (lucperkins) [reply modules/programs/buf.nix:20]: That's a fair question. There will probably be more configurability in the future but if this doesn't add much value now, I'm happy to close this and take it as a learning exercise.

**Comment by lucperkins** [comment]
This package is kind of pointless for now. I'll reopen this in the future if it makes more sense.

**Pull request closed without merging**